src/Tools/AnalyzerRunner/Program.cs

```diff
@@ -0,0 +1,480 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Diagnostics.Telemetry;
+using Microsoft.CodeAnalysis.MSBuild;
+using File = System.IO.File;
+using Path = System.IO.Path;
+
+namespace AnalyzerRunner
+{
+    /// <summary>
+    /// AnalyzerRunner is a tool that will analyze a solution, find diagnostics in it and will print out the number of
+    /// diagnostics it could find. This is useful to easily test performance without having the overhead of visual
+    /// studio running.
+    /// </summary>
+    class Program
+    {
+        public static async Task Main(string[] args)
+        {
+            var options = Options.Create(args);
+            if (options == null)
+            {
+                return;
+            }
+
+            CancellationTokenSource cts = new CancellationTokenSource();
+            Console.CancelKeyPress +=
+                (sender, e) =>
+                {
+                    e.Cancel = true;
+                    cts.Cancel();
+                };
+
+            var analyzerAssemblies = GetAnalyzerAssemblies(options.AnalyzerPath);
+            var analyzers = GetAllAnalyzers(analyzerAssemblies);
+            analyzers = FilterAnalyzers(analyzers, options).ToImmutableArray();
+
+            if (analyzers.Length == 0)
+            {
+                Utilities.WriteLine("No analyzers found", ConsoleColor.Red);
+                Utilities.PrintHelp();
+                return;
+            }
+            var cancellationToken = cts.Token;
+
+            Stopwatch stopwatch = Stopwatch.StartNew();
+            using (MSBuildWorkspace workspace = MSBuildWorkspace.Create())
+            {
+                Solution solution = await workspace.OpenSolutionAsync(options.SolutionPath, cancellationToken).ConfigureAwait(false);
```

Comment:
we probably want to make sure solution is all loaded successfully. I think we should get Compilation of each Project and call compilation.GetDiagnostics() to make sure it return 0 diagnostics.

Reply:
thinking little bit more, that might be cumbersome to run every single time especially if one is trying to measure and improve iteratively. it will be nice to have an option to verify MSBuildWorkspace.OpenSolutionAsync has worked properly though. otherwise, we have no idea whether solution we have here is broken or valid.

Reply:
why should not we proceed with broken projects?

---
In reply to: [149858404](https://github.com/dotnet/roslyn/pull/23087#discussion_r149858404) [](ancestors = 149858404)

Reply:
I am not saying we shouldn't, I am saying, if we did, we should let users know that.

let's say, you are testing diagnostic analyzer that report diagnostic if a type return Task and not use ConfigureAwait. you run tool analyzer roslyn.sln and you get 0 diagnostics back taking 2 min.

can you tell, whether it actually ran analyzers and report 0 diagnostics, or it returned 0 diagnostics because Roslyn.sln wasn't loaded properly (many reasons, nuget restore not happened, dll locked up and etc) so, all Task becomes error type so, analyzer actually didn't run properly and at the end, you get 0 diagnostics.

and we wrongly believe, this analyzer is either suddenly improved, or it is fast.